src/js/player.js

```diff
@@ -3023,14 +3023,62 @@ class Player extends Component {
    * continue consuming media while they interact with other content sites, or
    * applications on their device.
    *
+   * This can use document picture-in-picture or element picture in picture
+   *
+   * Set `enableDocumentPictureInPicture` to `true` to use docPiP on a supported browser
+   * Else set `disablePictureInPicture` to `false` to disable elPiP on a supported browser
+   *
+   *
    * @see [Spec]{@link https://wicg.github.io/picture-in-picture}
+   * @see [Spec]{@link https://wicg.github.io/document-picture-in-picture/}
    *
    * @fires Player#enterpictureinpicture
    *
    * @return {Promise}
    *         A promise with a Picture-in-Picture window.
    */
   requestPictureInPicture() {
+    if (this.options_.enableDocumentPictureInPicture && 'documentPictureInPicture' in window) {
+      const pipContainer = document.createElement(this.el().tagName);
+
+      pipContainer.classList = this.el().classList;
+      pipContainer.classList.add('vjs-pip-container');
+      if (this.posterImage) {
+        pipContainer.appendChild(this.posterImage.el().cloneNode(true));
+      }
+      if (this.titleBar) {
+        pipContainer.appendChild(this.titleBar.el().cloneNode(true));
+      }
+      pipContainer.appendChild(Dom.createEl('p', { className: 'vjs-pip-text' }, {}, this.localize('Playing in picture-in-picture')));
+
+      return new Promise((resolve, reject) => {
+        window.documentPictureInPicture.requestWindow({
+          // The aspect ratio won't be correct, Chrome bug https://crbug.com/1407629
+          initialAspectRatio: this.videoWidth() / this.videoHeight(),
+          copyStyleSheets: true
+        }).then(pipWindow => {
+          this.el_.parentNode.insertBefore(pipContainer, this.el_);
+
+          pipWindow.document.title = 'Video.js';
+          pipWindow.document.body.append(this.el_);
+          pipWindow.document.body.classList.add('vjs-pip-window');
+
+          this.player_.isInPictureInPicture(true);
+          this.player_.trigger('enterpictureinpicture');
+
+          // Listen for the PiP closing event to move the video back.
+          pipWindow.addEventListener('unload', (event) => {
+            const pipVideo = event.target.querySelector('.video-js');
+
+            pipContainer.replaceWith(pipVideo);
+            this.player_.isInPictureInPicture(false);
+            this.player_.trigger('leavepictureinpicture');
+          });
+
+          resolve(pipWindow);
+        }).catch(reject);
```

Comment:
question: Rather than `resolve` and `reject` here, would it make sense to return the promise returned by `window.documentPictureInPicture.requestWindow` above? I guess it's not clear to me what the purpose is of this outer promise.

Reply:
Ah yes I should return the window here instead.